tests/name_constraints/generate.py

```diff
@@ -0,0 +1,282 @@
+    ])
+
+    # ee
+    builder = x509.CertificateBuilder()
+    builder = builder.subject_name(x509.Name(
+        ([x509.NameAttribute(NameOID.COMMON_NAME, subject_common_name)] if subject_common_name else []) +
+        [x509.NameAttribute(NameOID.ORGANIZATION_NAME, test_name) ] +
+        extra_subject_names
+    ))
+    builder = builder.issuer_name(issuer_name)
+
+    builder = builder.not_valid_before(not_before)
+    builder = builder.not_valid_after(not_after)
+    builder = builder.serial_number(x509.random_serial_number())
+    builder = builder.public_key(public_key)
+    if sans:
+        builder = builder.add_extension(
+            x509.SubjectAlternativeName(sans),
+            critical=False
+        )
+    builder = builder.add_extension(
+        x509.BasicConstraints(ca=False, path_length=None), critical=True,
+    )
+    certificate = builder.sign(
+        private_key=issuer_private_key,
+        algorithm=hashes.SHA256(),
+        backend=default_backend(),
+    )
+
+    with open(test_name + '.ee.der', 'wb') as f:
+        f.write(certificate.public_bytes(Encoding.DER))
+
+    # issuer
+    builder = x509.CertificateBuilder()
+    builder = builder.subject_name(issuer_name)
+    builder = builder.issuer_name(issuer_name)
+    builder = builder.not_valid_before(not_before)
+    builder = builder.not_valid_after(not_after)
+    builder = builder.serial_number(x509.random_serial_number())
+    builder = builder.public_key(issuer_public_key)
+    builder = builder.add_extension(
+        x509.BasicConstraints(ca=True, path_length=None), critical=True,
+    )
+    if permitted_subtrees or excluded_subtrees:
+        builder = builder.add_extension(
+            x509.NameConstraints(permitted_subtrees, excluded_subtrees),
+            critical=True
+        )
+
+    certificate = builder.sign(
+        private_key=issuer_private_key,
+        algorithm=hashes.SHA256(),
+        backend=default_backend()
+    )
+
+    with open(test_name + '.ca.der', 'wb') as f:
+        f.write(certificate.public_bytes(Encoding.DER))
+
+    if expected_error is None:
+        expected = 'Ok(())'
+    else:
+        expected = 'Err(webpki::Error::' + expected_error + ')'
+
+    valid_names = ', '.join('"' + name + '"' for name in valid_names)
+    invalid_names = ', '.join('"' + name + '"' for name in invalid_names)
+
+    print("""
+#[test]
+#[cfg(feature = "alloc")]
+fn %(test_name)s() {
+    let ee = include_bytes!("name_constraints/%(test_name)s.ee.der");
+    let ca = include_bytes!("name_constraints/%(test_name)s.ca.der");
+    assert_eq!(
+        check_cert(ee, ca, &[%(valid_names)s], &[%(invalid_names)s]),
+        %(expected)s
+    );
+}""" % locals(), file=output)
+
+top = open('../name_constraints.rs', 'r').readlines()
+top = top[:top.index('// DO NOT EDIT BELOW: generated by name_constraints/generate.py\n')+1]
+output = open('../name_constraints.rs', 'w')
+for l in top:
+    output.write(l)
+
+name_constraints_test(
```

Comment:
Are the test cases here some existing body of tests, or did you make these up? What was your strategy for doing so?

Reply:
I made them up, partially from clauses in [RFC5280 4.2.1.10](https://www.rfc-editor.org/rfc/rfc5280#section-4.2.1.10) and partially from observing the existing code's behaviour